toolbox/oncall_alert/main.go

```diff
@@ -265,31 +286,64 @@ func getProwResult(target string) (*ProwResult, error) {
 	return &prowResult, nil
 }
 
-// Add "do-not-merge/post-submit" labels to all PRs in protected repo towards protected branch
-func blockPRs() {
-	options := github.PullRequestListOptions{
-		State: "open",
-		Base:  *protectedBranch,
+func processProwResult(job *jobStatus, runNo int, prowResult *ProwResult) *failure {
+	if *catchFlakesByRun {
+		resultSHA := prowResult.Metadata.RepoCommit
+		if flakeStatPtr, exists := job.rerunJobStats[resultSHA]; exists {
+			flakeStatPtr.TotalRerun++
+			if !prowResult.Passed {
+				flakeStatPtr.Failures++
+			}
+			if flakeStatPtr.TotalRerun == *numRerun {
+				log.Printf("All reruns on job [%s] at sha [%s] have finished\n", job.name, resultSHA)
+				if err := recordFlakeStatToGCS(job, *flakeStatPtr); err != nil {
+					log.Printf("Failed to write flakeStat to GCS: %v\n", err)
+				}
+				// delete resultSHA from job.rerunJobStats since all reruns have finished
+				delete(job.rerunJobStats, resultSHA)
+			}
+		} else { // no reruns exist on this SHA
+			if !prowResult.Passed {
+				log.Printf("Starting new rerun task on job [%s] at sha [%s]\n", job.name, resultSHA)
+				job.rerunJobStats[resultSHA] = &FlakeStat{
+					TestName:           job.name,
+					SHA:                resultSHA,
+					ParentJobTimeStamp: prowResult.TimeStamp,
+				}
+				if err := rerun(job, runNo); err != nil {
+					log.Printf("failed when starting reruns on [%s]: %v\n", job.name, err)
+				}
+			}
+		}
 	}
-
-	log.Printf("Adding [%s] to PRs in %s", doNotMergeLabel, *protectedRepo)
-	if err := githubClnt.AddLabelToPRs(options, *protectedRepo, doNotMergeLabel); err != nil {
-		log.Printf("Failed to add label to PRs: %v", err)
-		return
+	if !prowResult.Passed {
+		return &failure{
+			jobName: job.name,
+			runNo:   runNo,
+		}
 	}
-	log.Printf("Blocked auto-merge in %s, base: %s", *protectedRepo, *protectedBranch)
+	return nil
 }
 
-// remove "do-not-merge/post-submit" labels to all PRs in protected repo towards protected branch
-func unBlockPRs() {
-	options := github.PullRequestListOptions{
-		State: "open",
-		Base:  *protectedBranch,
+func recordFlakeStatToGCS(job *jobStatus, newFlakeStat FlakeStat) error {
+	flakeStatsFile := filepath.Join(job.name, flakeStatsJSON)
+	flakeStatsString, err := gcsClient.GetFileFromGCSString(*gcsBucket, flakeStatsFile)
+	if err != nil {
+		return err
 	}
-
-	if err := githubClnt.RemoveLabelFromPRs(options, *protectedRepo, doNotMergeLabel); err != nil {
-		log.Printf("Failed to remove label to PRs: %v", err)
-		return
+	flakeStats, err := DeserializeFlakeStats(flakeStatsString)
+	if err != nil {
+		return err
+	}
+	flakeStats = append(flakeStats, newFlakeStat)
```

Comment:
You cannot use GCS storage for this, as you cannot guarantee consistenty. You want to use storage that supports transaction. Maybe bigquery ? Have you looked at the flaky detection that k8s is using ? I think they already do all this anyways.

Reply:
Understand. After talking with @hklai I think it is a good idea to work on the storage in a separate PR, so that we could have the rerun/retest running first, and prow shows these rerun too. In the meantime, I will update the design doc to have a design on the durable transactional storage, and we can carry on the discussion there.